### PR TITLE
Address Manager

### PR DESCRIPTION
**Pull request by ConnorBryan**
### Description
Adds functionality related to an "Address Book" to MyCrypto:
* An "Address Book" tab after logging in where new <Address: Label> pairs can be entered.
address-book-label

* The `AccountInfo` component on the right side of the account now displays one of three states based on whether the user has entered a label for their own address:

1. No label, a button shows prompting the quick creation.

add-label-account-info

2. A label exists, shows an edit icon as well as the label.

screen shot 2018-04-21 at 1 37 42 pm

3. The label is being edited from `AccountInfo`

screen shot 2018-04-21 at 1 38 10 pm

* The `AddressField` component now displays a dropdown when the value contains two or more characters. This dropdown is navigable via arrow keys and will enter the address into the field on selection.

screen shot 2018-04-21 at 1 40 10 pm

* If the value in the `AddressField` component matches a label, the label for the input field will change.

screen shot 2018-04-21 at 1 42 09 pm

* Addresses in the `DeterministicWalletModal` now display a label on top if one exists.

deterministic-wallet-modal

### Changes
* Created a new tab in `Tabs/SendTransaction`: `AddressBook`
* Created new components: AddressBookTable, AddressBookTableRow, AddressLabel
* Created an addition to `AddressFieldFactory` that acts as a dropdown
* Implemented a fully tested `addressBook` reducer that adds labels to both LocalStorage and the redux state
* Added new tested selectors to get pieces of `addressBook`

### Steps to Test
1. Log into an account
2. Navigate to the `Address Book` tab
3. Click the "+" button next to the Add new address/label row
* The `Add new address` field should be selected
4. Enter an invalid address
5. Click the "+" button again
* The `Add new address` field should be selected again
6. Enter a valid Ethereum address
7. Enter the label "Minni"
8. Press <ENTER>
* A new row should be added to the table with the entered Address/Minni pair.
9. Click on the Minni field, change it to Minnie, press <ENTER>
* The field should no longer be highlighted.
10. On the `AccountInfo` component, press the `Add Label` button.
11. In the input, enter `Mine` and press enter.
12. The label should now show `Mine`
13. The `AddressBook` table in the active tab should now have a new entry for `Mine`.
14. Press the `X` button next to `Minnie` label.
* It should be removed.
15. Navigate to the `Send Ether & Tokens` tab.
16. Enter `Mi` into the To field.
* A dropdown should appear showing `Mine` with the address.
17. Press the <DOWN ARROW> key, then the <ENTER> key.
18. The address for `Mine` should populate the field and the label should read (check) Mine

### CODE REVIEW

File: common/components/AddressBookTable.tsx [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+          {rows.map(this.makeLabelRow)}
+        </tbody>
+      </table>
+    );
+  }
+
+  private handleSave = (addressToLabel: AddressLabelPair) => {
+    const { label, address } = addressToLabel;
+
+    this.props.addLabelForAddress({ label, address });
```
Comment (HenryNguyen5) [inline common/components/AddressBookTable.tsx]: No need to destructure here, just pass the arg along

File: common/components/AddressBookTable.tsx [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+  private handleKeyDown = (e: React.KeyboardEvent<HTMLTableElement>) => {
+    if (e.keyCode === KeyCodes.ENTER) {
+      this.handleAddEntry();
+    }
+  };
+
+  private setEditingRow = (editingRow: number | null) => this.setState({ editingRow });
+
+  private removeEntry = (address: string) => this.props.removeLabelForAddress(address);
```
Comment (HenryNguyen5) [inline common/components/AddressBookTable.tsx]: The two above methods dont really seem needed, can just call the associated functionality directly

File: common/selectors/transactions.ts [hunk 3]
```diff
@@ -28,3 +28,7 @@ export function getRecentWalletTransactions(state: AppState): SavedTransaction[]
     return [];
   }
 }
+
+export function getTransactionToRaw(state: AppState) {
```
Comment (HenryNguyen5) [inline common/selectors/transactions.ts]: Almost duplicate of `getTo` in `common/selectors/transaction/fields`, you can just extend `getTo` to return `raw` in another function

File: common/selectors/transactions.ts [hunk 4]
```diff
@@ -28,3 +28,7 @@ export function getRecentWalletTransactions(state: AppState): SavedTransaction[]
...
+  return state.transaction.fields.to.raw;
```
Comment (HenryNguyen5) [inline common/selectors/transactions.ts]: Also the selectors in this file are for `state.transactions` not `state.transaction`

File: common/components/BalanceSidebar/AccountInfo.tsx [hunk 5]
```diff
@@ -156,16 +160,16 @@ class AccountInfo extends React.Component<Props, State> {
                 onClick={this.toggleShowLongBalance}
               >
                 <UnitDisplay
-                  value={balance.wei}
+                  value={etherBalance}
```
Comment (HenryNguyen5) [inline common/components/BalanceSidebar/AccountInfo.tsx]: Unrelated code refactor while I was cleaning up the types in this component. My bad.

File: common/components/AddressFieldFactory/AddressFieldDropdown.scss [hunk 6]
```diff
@@ -0,0 +1,42 @@
+@import 'common/sass/variables';
+@import 'common/sass/variables/colors';
+@import 'common/sass/variables/spacing';
+
+.AddressFieldDropdown {
+  position: absolute;
+  top: calc(100% - 16px);
```
[the remark this answers is not in the document]
Comment (wbobeirne) [reply common/components/AddressFieldFactory/AddressFieldDropdown.scss]: This should also be the width of the input, so it should be a in a position: relative div around the input, and have `left: 0; width: 100%;`

File: common/components/AddressFieldFactory/AddressFieldDropdown.scss [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+    padding: $space-md $space;
+    display: flex;
+    align-items: center;
+
+    &:hover {
+      cursor: pointer;
+      background: $gray-lightest;
+    }
+
+    strong:first-of-type {
```
Comment (wbobeirne) [inline common/components/AddressFieldFactory/AddressFieldDropdown.scss]: Prefer classnames.

File: common/components/AddressFieldFactory/AddressFieldDropdown.scss [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      width: $space-xl * 6;
```
Comment (wbobeirne) [inline common/components/AddressFieldFactory/AddressFieldDropdown.scss]: Rather than give labels a fixed width, which can look weird if they're all short or you only have one, I'd suggest doing

(Identicon) **Label** _address_

So that it can just be left aligned as normal.

File: common/components/AddressFieldFactory/AddressFieldDropdown.tsx [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+      switch (e.keyCode) {
+        case KeyCodes.ENTER:
+          return this.handleEnterKeyDown();
+        case KeyCodes.UP_ARROW:
+          return this.handleUpArrowKeyDown();
+        case KeyCodes.DOWN_ARROW:
+          return this.handleDownArrowKeyDown();
+        default:
+          return;
+      }
```
Comment (wbobeirne) [inline common/components/AddressFieldFactory/AddressFieldDropdown.tsx]: This is dope, make sure you also preventDefault() if it's up / down / enter.

File: common/components/AddressField.tsx [hunk 10]
```diff
@@ -14,23 +14,37 @@ interface Props {
...
-    )}
+    withProps={({ currentLabel, currentTo, isValid, onChange, onFocus, onBlur, readOnly }) => {
+      let label;
+
+      if (currentLabel && !isSelfAddress) {
+        label = (
+          <span>
+            <i className="fa fa-check" /> {currentLabel}
+          </span>
+        );
```
Comment (wbobeirne) [inline common/components/AddressField.tsx]: I'm not wild about changing the label for accessibility reasons. I'd prefer this be some green text below the input or something. Ideally this can be unified with the ENS resolver, and share the same styles. That one could use some UI love.

File: common/components/AddressFieldFactory/AddressFieldFactory.tsx [hunk 11]
```diff
@@ -33,16 +48,36 @@ class AddressFieldFactoryClass extends React.Component<Props> {
...
+    if (this.goingToBlur) {
+      window.clearTimeout(this.goingToBlur);
+    }
+  }
+
   public render() {
+    const { isFocused } = this.state;
+
+    // Allow the click event on a dropdown item to process before hiding the dropdown.
+    const onBlur = () => (this.goingToBlur = window.setTimeout(this.blur, 150));
```
Comment (wbobeirne) [inline common/components/AddressFieldFactory/AddressFieldFactory.tsx]: This should be a class function, otherwise it'll always cause a re-render since it's a new function each time.

File: shared/keycodes.ts [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+  [key: string]: number;
+}
+
+const KeyCodes: KeyMapping = {
+  ENTER: 13,
+  UP_ARROW: 38,
+  DOWN_ARROW: 40
+};
+
+export default KeyCodes;
```
Comment (wbobeirne) [inline shared/keycodes.ts]: React provides some more readable keyboard event attributes that don't require the mapping https://reactjs.org/docs/events.html#keyboard-events

File: common/components/AddressBookTable.tsx [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+            </td>
+          </tr>
+          {rows.map(this.makeLabelRow)}
+        </tbody>
+      </table>
+    );
+  }
+
+  private handleSave = (addressToLabel: AddressLabelPair) => {
+    const { label, address } = addressToLabel;
```
[the remark this answers is not in the document]
Comment (wbobeirne) [reply common/components/AddressBookTable.tsx]: Likewise, we should not allow for duplicate labels. Perhaps this should be handled as a saga.

**Comment by dternyak** [comment]
@wbobeirne I like the touchups quite a bit! Would be great to get those changes in.

File: common/components/AddressBookTable.tsx (hunk 1 above)
Comment (ConnorBryan) [reply common/components/AddressBookTable.tsx]: 👍

File: common/components/AddressBookTable.tsx (hunk 2 above)
Comment (ConnorBryan) [reply common/components/AddressBookTable.tsx]: I'll agree on `removeEntry`, but I like keeping `setState` calls in their own method as a personal quirk. I don't think it causes harm, and it improves readability.

:+1: on `removeEntry`

File: common/selectors/transactions.ts (hunk 4 above)
Comment (ConnorBryan) [reply common/selectors/transactions.ts]: :+1:

File: common/selectors/transactions.ts (hunk 3 above)
Comment (ConnorBryan) [reply common/selectors/transactions.ts]: :+1:

File: common/components/BalanceSidebar/AccountInfo.tsx (hunk 5 above)
Comment (ConnorBryan) [reply common/components/BalanceSidebar/AccountInfo.tsx]: 👻

File: common/components/AddressFieldFactory/AddressFieldFactory.tsx (hunk 11 above)
Comment (ConnorBryan) [reply common/components/AddressFieldFactory/AddressFieldFactory.tsx]: :+1:

File: common/components/AddressBookTable.tsx [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+          <tr>
+            <td>
+              <div className="AddressBookTable-cell">
+                <Input
+                  placeholder="New address"
+                  value={temporaryAddress}
+                  onChange={this.setTemporaryAddress}
+                  setInnerRef={this.setAddressInputRef}
+                />
+                <Identicon address={temporaryAddress} />
```
[the remark this answers is not in the document]
Comment (ConnorBryan) [reply common/components/AddressBookTable.tsx]: :+1:

File: shared/keycodes.ts (hunk 12 above)
Comment (ConnorBryan) [reply shared/keycodes.ts]: :+1:

File: common/components/AddressFieldFactory/AddressFieldDropdown.tsx (hunk 9 above)
Comment (ConnorBryan) [reply common/components/AddressFieldFactory/AddressFieldDropdown.tsx]: :+1:

File: common/components/AddressBookTableRow.tsx [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+          </div>
+        </td>
+        <td>
+          <div className="AddressBookTable-cell">
+            <button
+              title="Remove this entry"
+              className="btn btn-sm  btn-danger"
+              onClick={onRemoveClick}
+            >
+              <i className="fa fa-close" />
```
[the remark this answers is not in the document]
Comment (ConnorBryan) [reply common/components/AddressBookTableRow.tsx]: :+1:

File: common/components/AddressBookTable.tsx [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+                  value={temporaryLabel}
+                  onChange={this.setTemporaryLabel}
+                  setInnerRef={this.setLabelInputRef}
+                />
+              </div>
+            </td>
+            <td>
+              <div className="AddressBookTable-cell">
+                <button className="btn btn-sm btn-default" onClick={this.handleAddEntry}>
+                  <i className="fa fa-plus" />
```
[the remark this answers is not in the document]
Comment (ConnorBryan) [reply common/components/AddressBookTable.tsx]: :+1:

File: common/components/AddressFieldFactory/AddressFieldDropdown.scss (hunk 6 above)
[the remark this answers is not in the document]
Comment (ConnorBryan) [reply common/components/AddressFieldFactory/AddressFieldDropdown.scss]: Top comment: :+1: , though I just used `$space` since the margin-bottom is set in `input-group-input`.

Bottom comment: 👍

File: common/components/AddressFieldFactory/AddressFieldDropdown.scss (hunk 7 above)
Comment (ConnorBryan) [reply common/components/AddressFieldFactory/AddressFieldDropdown.scss]: :+1:

File: common/components/AddressFieldFactory/AddressFieldDropdown.scss (hunk 8 above)
Comment (ConnorBryan) [reply common/components/AddressFieldFactory/AddressFieldDropdown.scss]: :+1:

File: common/components/AddressField.tsx (hunk 10 above)
Comment (ConnorBryan) [reply common/components/AddressField.tsx]: :+1:

File: common/translations/lang/en.json [hunk 17]
```diff
@@ -1,6 +1,15 @@
 {
   "code": "en",
   "data": {
+    "ADDRESS": "Address",
+    "LABEL": "Label",
+    "INVALID_LABEL_LENGTH": "An address label must be between 2 and 20 characters.",
```
Comment (james-prado) [inline common/translations/lang/en.json]: Why are we imposing a character limit for labels? If responsiveness is an issue, we can always truncate the labels.

Comment (james-prado) [reply common/translations/lang/en.json]: Also, addresses should be truncated on smaller screens. Personally I like the way metamask does it, they truncate the middle section of an address so you can verify the first 3 and last 3 digits of an address.

File: common/components/AddressFieldFactory/AddressFieldDropdown.scss [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+.AddressFieldDropdown {
+  position: absolute;
+  top: calc(100% - 16px);
+  border: 1px solid $gray-lighter;
+  border-top: none;
+  background: $component-active-color;
+  z-index: 1;
+  list-style-type: none;
+  margin: 0;
+  padding: 0;
```
[the remark this answers is not in the document]
Comment (james-prado) [reply common/components/AddressFieldFactory/AddressFieldDropdown.scss]: I'm actually the opposite, I think that dropdown style throughout the site should remain consistent. Maybe this would be a good time to change how we do dropdowns / input styles?

All the other dropdowns just follow the simple input style, but I'm definitely open to updating those. I'm also not opposed to the drop-shadow, but I do think dropdowns should match the input style.

File: common/components/AddressFieldFactory/AddressFieldDropdown.tsx [hunk 19]
```diff
@@ -0,0 +1,157 @@
...
+
+      return (
+        <li
+          key={address}
+          className={className}
+          onClick={() => this.props.setCurrentTo(address)}
+          role="option"
+          title={`${translateRaw('SEND_TO')}${label}`}
+        >
+          <Identicon address={address} size="2rem" />
```
Comment (james-prado) [inline common/components/AddressFieldFactory/AddressFieldDropdown.tsx]: On smaller screens the EthBlockies icon still gets squished 
screen shot 2018-05-01 at 3 31 38 pm

File: common/components/AddressFieldFactory/AddressFieldFactory.tsx [hunk 20]
```diff
@@ -15,16 +16,28 @@ interface OwnProps {
...
+interface State {
+  isFocused: boolean;
+}
+
 export interface CallbackProps {
   isValid: boolean;
   readOnly: boolean;
   currentTo: ICurrentTo;
   onChange(ev: React.FormEvent<HTMLInputElement>): void;
+  onFocus(ev: React.FormEvent<HTMLInputElement>): void;
```
Comment (james-prado) [inline common/components/AddressFieldFactory/AddressFieldFactory.tsx]: The onFocus for the address input should be updated so it isn't `invalid` when searching for an address in your address book. 
screen shot 2018-05-01 at 3 33 27 pm


This occurs after:
- Begin searching for a valid address
- leave the input (`onBlur()`)
- focus the input (`onFocus()`)

Comment (james-prado) [reply common/components/AddressFieldFactory/AddressFieldFactory.tsx]: Also, if you start typing an address label that doesn't exist, it should display an error, something like 'That address doesn't exist in your address book'.

File: common/components/AddressBookTable.scss [hunk 21]
```diff
@@ -0,0 +1,85 @@
+@import 'common/sass/variables/spacing';
+@import 'common/sass/variables/colors';
+
+.AddressBookTable {
```
Comment (james-prado) [inline common/components/AddressBookTable.scss]: The Address book is still lacking some support for smaller screens.

screen shot 2018-05-01 at 3 42 22 pm

File: common/components/AddressFieldFactory/AddressFieldDropdown.scss (hunk 18 above)
[the remark this answers is not in the document]
Comment (ConnorBryan) [reply common/components/AddressFieldFactory/AddressFieldDropdown.scss]: That may be the case, but the scope on this has already exploded. This would probably be better as a future story.

File: common/components/AddressBookTable.scss (hunk 21 above)
Comment (ConnorBryan) [reply common/components/AddressBookTable.scss]: Oh, that must be right before it collapses. Should be a quick fix.

File: common/components/AddressFieldFactory/AddressFieldFactory.tsx (hunk 20 above)
Comment (ConnorBryan) [reply common/components/AddressFieldFactory/AddressFieldFactory.tsx]: Agree on the bottom, but I was told that an I out should show invalid if it’s invalid. Even if you’re searching for a label, at that moment it’s not actually valid.

Clarification, @dternyak ?

File: common/translations/lang/en.json (hunk 17 above)
Comment (ConnorBryan) [reply common/translations/lang/en.json]: I imposed the limit because I was asked to by will. Truncation is fine, but again, in the interest of avoiding story bloat that more advanced functionality might be better as a different story.

File: common/components/AddressFieldFactory/AddressFieldFactory.tsx (hunk 20 above)
Comment (dternyak) [reply common/components/AddressFieldFactory/AddressFieldFactory.tsx]: @ConnorBryan The input that I mentioned that should have invalid styling when invalid is the address book address input, not the To Address input.

Comment (ConnorBryan) [reply common/components/AddressFieldFactory/AddressFieldFactory.tsx]: @dternyak

Okay! I’ll get this fixed up when I get home from the hospital later.

File: common/components/AddressFieldFactory/AddressFieldDropdown.tsx (hunk 19 above)
Comment (ConnorBryan) [reply common/components/AddressFieldFactory/AddressFieldDropdown.tsx]: Whoops — I need to wrap those in containers with a minimum width. :+1:

**Comment by ConnorBryan** [comment]
@dternyak

I like the idea of incorporating validation to input factory in a future story. For now I can spend some time making the notification that displays in submit appear next to the label, aligned right, colored $beand-danger. Would that be acceptable?

**Comment by dternyak** [comment]
Agreed, the input/error factory should be part of a future story.

r/e alignment: I would checkout what we do on the swap view when errors are exceed max or are less than the minimum. I think this is reasonable, although I'm sure improvements can be had all around.

screen shot 2018-05-01 at 3 29 20 pm

**Comment by james-prado** [comment]
@dternyak Yea, I think with all the updates since the last time all the inputs got some attention it might be time to clean them up again. I'll open up a new issue for updating and standardizing all our input types like inline dropdown, searchable dropdown, inline loading msgs, etc... Also standardizing error msgs and how they're displayed will be covered there too.

File: common/components/AddressFieldFactory/AddressFieldDropdown.tsx (hunk 19 above)
Comment (ConnorBryan) [reply common/components/AddressFieldFactory/AddressFieldDropdown.tsx]: Fixed 👍 

screen shot 2018-05-02 at 3 13 41 pm

File: common/translations/lang/en.json (hunk 17 above)
Comment (ConnorBryan) [reply common/translations/lang/en.json]: Fixed: Updated to 50 and introduced truncation in the various spots that display the label.

**Comment by ConnorBryan** [comment]
# AddressFieldDropdown

## AddressFieldDropdown, desktop, match exists
screen shot 2018-05-03 at 11 24 22 pm

_Note: The dropdown can be interacted with via the up and down arrow keys, as well as `Enter` for selection._

## AddressFieldDropdown, desktop, no match exists
screen shot 2018-05-03 at 11 27 13 pm

## AddressFieldDropdown, desktop, invalid address entry
screen shot 2018-05-03 at 11 28 50 pm

_Note: Entering any text other than '0x' as the first characters of the query will disable validation highlighting, as it is assumed the user is searching for a label._

## AddressFieldDropdown, mobile, match exists
screen shot 2018-05-03 at 11 25 43 pm

## AddressFieldDropdown, mobile, no match exists
screen shot 2018-05-03 at 11 27 54 pm


# AddressBookTable

## AddressBookTable, desktop, default view
screen shot 2018-05-03 at 11 48 52 pm

## AddressBookTable, desktop, with errors
screen shot 2018-05-03 at 11 49 50 pm

## AddressBookTable, desktop, submitting with errors
screen shot 2018-05-03 at 11 50 33 pm

## AddressBookTable, desktop, error on row
screen shot 2018-05-03 at 11 51 11 pm

## AddressBookTable, mobile, default view
screen shot 2018-05-03 at 11 52 34 pm

## AddressBookTable, mobile, with errors
screen shot 2018-05-03 at 11 53 17 pm

## AddressBookTable, mobile, submitting with errors
screen shot 2018-05-03 at 11 53 52 pm

## AddressBookTable, mobile, error on row
screen shot 2018-05-03 at 11 54 46 pm


# AccountAddress

## AccountAddress, default view
screen shot 2018-05-03 at 11 55 24 pm

## AccountAddress, editing view
screen shot 2018-05-03 at 11 56 24 pm

## AccountAddress, no label
screen shot 2018-05-03 at 11 56 55 pm

## AccountAddress, error
screen shot 2018-05-03 at 11 57 24 pm

## AccountAddress, submitting with error
screen shot 2018-05-03 at 11 58 03 pm

# Notes

Okay, I'm pretty sure I've addressed all comments. I've attached updated pictures of the various state. Please let me know any other changes you'd like made prior to merge.

@dternyak 
@HenryNguyen5 
@wbobeirne 
@james-prado

**Comment by james-prado** [comment]
@ConnorBryan The reason for changing the validation so it only occurs after the first time an input has blurred was to reduce the amount of error messages being displayed when a user hadn't even had a change to fill in the content for a given input. Ideally, the user should have a chance to fill in the content correctly before throwing errors at them. The only exception to this is if a user tries to submit a form without filling all the required inputs.

File: common/components/AddressBookTable.tsx [hunk 22]
```diff
@@ -0,0 +1,316 @@
...
+    const { addressTouched, addressBlurred, labelTouched, labelBlurred } = this.state;
+
+    // Classnames
+    const addressTouchedWithError = addressTouched && addressError;
+    const labelTouchedWithError = labelTouched && labelError;
+    const addressInputClassName = addressTouchedWithError ? 'invalid' : '';
+    const labelInputClassName = labelTouchedWithError ? 'invalid' : '';
+    const nonMobileTemporaryInputErrorClassName =
+      'AddressBookTable-row-error-temporary-input--non-mobile';
+    const nonMobileVisibleErrorClassName = `${nonMobileTemporaryInputErrorClassName}--visible`;
```
Comment (wbobeirne) [inline common/components/AddressBookTable.tsx]: For quick little modifiers like these, we have a precedent of using state classes like `is-visible` or `is-non-mobile` to save some line length.

File: common/components/AddressBookTableRow.tsx [hunk 23]
```diff
@@ -0,0 +1,149 @@
...
+            </div>
+          </div>
+          <div className="AddressBookTable-row-input">
+            <div className="AddressBookTable-row-input-wrapper">
+              <label htmlFor={labelName} className="AddressBookTable-row-input-wrapper-label">
+                {translate('LABEL')}
+              </label>
+              <Input
+                name={labelName}
+                title={`${translateRaw('EDIT_LABEL_FOR')}${address}`}
```
Comment (wbobeirne) [inline common/components/AddressBookTableRow.tsx]: There should be a space between these two.

File: common/components/AddressBookTableRow.tsx [hunk 24]
```diff
@@ -0,0 +1,149 @@
...
+              <label className="AddressBookTable-row-input-wrapper-error">
+                <div />
+                {labelError}
+              </label>
+            </div>
+          )}
+        </div>
+        {labelError && (
+          <div className="AddressBookTable-row AddressBookTable-row-error AddressBookTable-row-error--non-mobile">
+            <div />
```
Comment (wbobeirne) [inline common/components/AddressBookTableRow.tsx]: What're these empty divs for? If you're trying to make something display: block, better to do that in CSS.

File: common/components/AddressFieldFactory/AddressFieldDropdown.tsx [hunk 25]
```diff
@@ -0,0 +1,165 @@
...
+
+  public componentWillUnmount() {
+    window.removeEventListener('keydown', this.handleKeyDown);
+  }
+
+  public render() {
+    const { currentTo } = this.props;
+    const noMatchContent = currentTo.startsWith('0x') ? null : (
+      <li className="AddressFieldDropdown-dropdown-item AddressFieldDropdown-dropdown-item-no-match">
+        <i className="fa fa-warning" /> {translate('NO_LABEL_FOUND_CONTAINING')}"{currentTo}".
```
Comment (wbobeirne) [inline common/components/AddressFieldFactory/AddressFieldDropdown.tsx]: Should be a space between translate and quotes.

Comment (wbobeirne) [reply common/components/AddressFieldFactory/AddressFieldDropdown.tsx]: For context, I think that some of the translations have trailing spaces, but I don't want to rely on that in case someone forgets to add one when doing the translations in other languages. Better to do it once for all cases.

File: common/components/AddressBookTableRow.tsx (hunk 23 above)
Comment (ConnorBryan) [reply common/components/AddressBookTableRow.tsx]: :+1:

File: common/components/AddressBookTableRow.tsx (hunk 24 above)
Comment (ConnorBryan) [reply common/components/AddressBookTableRow.tsx]: :+1:

File: common/components/AddressFieldFactory/AddressFieldDropdown.tsx (hunk 25 above)
Comment (ConnorBryan) [reply common/components/AddressFieldFactory/AddressFieldDropdown.tsx]: :+1:

File: common/components/AddressBookTable.tsx (hunk 22 above)
Comment (ConnorBryan) [reply common/components/AddressBookTable.tsx]: @wbobeirne Can you point to an example?

Comment (wbobeirne) [reply common/components/AddressBookTable.tsx]: https://github.com/MyCryptoHQ/MyCrypto/blob/aa0f9cd455aab2bd7134036fee858de4ae10028c/common/components/NavigationLink.tsx#L28 is an example, here's the SuitCSS guide on the matter https://github.com/suitcss/suit/blob/master/doc/naming-conventions.md#is-stateOfComponent

**Comment by avanimiriyala** [comment]
Hey Team, I have one small recommendation here in the design. In the Address Book, is it possible to move the wallet avatars to the left instead of keeping them in the middle? Not a huge UX issue, but more of a design polish thing. See attached mockup I made :) 
address-book
